src/jit/stack.cc

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2018 wasmjit-omr project participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "stack.h"
+#include "infra/Assert.hpp"
+
+namespace wabt {
+namespace jit {
+
+void VirtualStack::Push(TR::IlValue* v) {
+  values_.push_back(v);
+}
+
+TR::IlValue* CheckResult(TR::IlValue* v) {
+  TR_ASSERT(v, "Attempt to pick/pop an invalid value on the stack");
+  return v;
+}
+
+TR::IlValue* VirtualStack::Pop() {
+  TR_ASSERT(values_.size() > 0, "Attempt to pop off an empty stack");
+
+  auto* v = values_.back();
+  values_.pop_back();
+
+  return CheckResult(v);
+}
+
+void VirtualStack::DropKeep(size_t drop_count, size_t keep_count) {
+  TR_ASSERT(drop_count + keep_count >= drop_count && values_.size() > drop_count + keep_count,
+            "Attempt to drop_keep beyond the end of the stack");
+
+  values_.erase(values_.end() - drop_count - keep_count, values_.end() - keep_count);
+}
+
+TR::IlValue* VirtualStack::Pick(size_t i) {
+  TR_ASSERT(i < values_.size(), "Attempt to pick beyond the end of the stack");
+
+  return CheckResult(*(values_.end() - i - 1));
+}
+
+TR::IlValue* VirtualStack::PickBottom(size_t i) {
+  TR_ASSERT(i < values_.size(), "Attempt to pick beyond the end of the stack");
+
+  return CheckResult(*(values_.begin() + i));
+}
+
+void VirtualStack::MergeInto(const VirtualStack* other, TR::IlBuilder* b) {
+  TR_ASSERT(values_.size() == other->values_.size(), "Attempt to merge divergent stacks");
+
+  auto this_it = values_.begin();
+  auto other_it = other->values_.begin();
+
+  while (this_it != values_.end()) {
```

Comment:
I've played around with this. As part of solving another problem you mentioned, I ended up adding stack offsets to the `TR_ASSERT` messages, however it seems like this change would make that more difficult since that would require keeping a separate counter for the stack offset instead of being able to use `this_it - values_.begin()`. In the interest of being able to do that easily, I'm leaning towards leaving this as-is.

Reply:
Makes sense. I agree the stack offset is useful to have around for debugging purposes.